### DISCUSSION TITLE
Numerical comparison in Wazuh Rules

### DISCUSSION THREAD

**Discussion by FriMeDev** [wazuh/wazuh#20696] [Q&A]
Hello, 

is it possible to simply compare numerical values by `>,>=,<,<=` in Wazuh Rules? I just want to fire an alert if a fields value is greater than number x. To do this with a regular expression is not ideal and I hope I am just missing something. 

Greetings

**Comment by Stuti3097** [comment]
Hi,

It's not possible to perform direct numerical comparisons like `>`, `<`, etc., within the Wazuh rule engine itself right now

However, you can work around this limitation by writing a standard rule that matches the event and triggers an active response script. In this custom script, you can read the latest alert from `/var/ossec/logs/alerts/alerts.json`, extract the numeric value, and apply your comparison logic (e.g., if value > 100). If the condition is met, the script can write a custom log entry to a separate log file.
Refer to https://documentation.wazuh.com/current/user-manual/capabilities/active-response/index.html

You can then configure Wazuh to monitor this custom log file using the `<localfile>` option in the configuration:

```xml
<localfile>
  <log_format>syslog</log_format>
  <location>/var/log/my_custom_alerts.log</location>
</localfile>
```

Reference:
[https://documentation.wazuh.com/current/user-manual/capabilities/log-data-collection/monitoring-log-files.html](https://documentation.wazuh.com/current/user-manual/capabilities/log-data-collection/monitoring-log-files.html)

Next, you would need to write a decoder and a rule to parse the new log and trigger an alert accordingly.

Reference:
[https://documentation.wazuh.com/current/user-manual/ruleset/index.html](https://documentation.wazuh.com/current/user-manual/ruleset/index.html)

**Comment by leonfullxr** [comment]
Hello,

Wazuh’s built-in syscheck/FIM decoder already extracts file sizes into the alert as syscheck.size_after (and syscheck.size_before on changes).

Here are some custom rules you can use to check whenever the file size goes over 100MB:
```
<group name="syscheck,local">
  <!-- >= 100 MB (100,000,000 bytes) -->
  <rule id="100017" level="8">
    <if_sid>550,554</if_sid>
    <field name="size" type="pcre2">^\d{9,}$</field>
    <description>FIM: file size ≥ 100 MB (size $(size) bytes): $(file)</description>
  </rule>

  <!-- >= 1 GB (1,000,000,000 bytes) -->
  <rule id="100018" level="12">
    <if_sid>100017</if_sid>
    <field name="size" type="pcre2">^\d{10,}$</field>
    <description>FIM: file size ≥ 1 GB (size $(size) bytes): $(file)</description>
  </rule>

  <!-- >= 1 TB (1,000,000,000,000 bytes) -->
  <rule id="100019" level="15">
    <if_sid>100018</if_sid>
    <field name="size" type="pcre2">^\d{13,}$</field>
    <description>FIM: file size ≥ 1 TB (size $(size) bytes): $(file)</description>
  </rule>
</group>
```
I have done some tests and it seems to be working as expected (see attached screenshot). The rule engine sees the FIM field name size (not size_after). Wazuh maps it to syscheck.size_after under the hood.
<img width="1916" height="951" alt="image" src="https://github.com/user-attachments/assets/a673b81d-ca7d-4714-b3c7-c19750f725cd" />


Here is the reference for creating custom FIM rules: https://documentation.wazuh.com/current/user-manual/capabilities/file-integrity/creating-custom-fim-rules.html#creating-custom-fim-rules

If you ever monitor multi-GB files, note the agent’s monitored file size limit (default ~1 GB). You can raise it (syscheck.file_max_size=4095). https://groups.google.com/g/wazuh/c/m3eitlmrnz0

If you prefer real numeric comparisons (no regex), create an alert in the Wazuh dashboard using Wazuh Query Language:
```
rule.id:(550 OR 554) AND syscheck.size_after>104857600
```
WQL supports the > operator for numeric fields, so this is clean and exact. https://documentation.wazuh.com/current/user-manual/wazuh-dashboard/queries.html

Reference: https://documentation.wazuh.com/current/user-manual/capabilities/file-integrity/basic-settings.html#record-file-attributes

Let me know if this helped
